README.md
<div align="center">
<br/><br/>
<h1>TimerList</h1>
<br/><br/>
Timerlist is a small module that allows you to have a reference of your timers.
<br/><br/>
<a href="https://travis-ci.org/fabioricali/timerlist" target="_blank"><img src="https://travis-ci.org/fabioricali/timerlist.svg?branch=master" title="Build Status"/></a>
<a href="https://coveralls.io/github/fabioricali/timerlist?branch=master" target="_blank"><img src="https://coveralls.io/repos/github/fabioricali/timerlist/badge.svg?branch=master" title="Coverage Status"/></a>
<a href="https://opensource.org/licenses/MIT" target="_blank"><img src="https://img.shields.io/badge/License-MIT-yellow.svg" title="License: MIT"/></a>
<img src="https://img.shields.io/badge/team-terrons-orange.svg" title="Team Terrons"/>
</div>

## Installation

### Node.js
```
npm install timerlist --save
```

## Example
```javascript
const TimerList = require('timerlist');
const timer = new TimerList();

// Adds a timeout timer
timer.setTimeout('my timer', ()=>{
    console.log('timer boom');
}, 2000);

// Adds an interval timer
timer.setInterval('my timer', ()=>{
    console.log('timer next');
}, 100);

// Passing params to listener
timer.setTimeout('my timer', (param)=>{
    console.log('a param', param);
}, 2000, 'Hello');

// Destroy timeout timer
timer.clearTimeout('my timeout');

// Destroy interval timer
timer.clearInterval('my interval');

// Destroy all timers
timer.clearAll();
```

### API Documentation
See <a href="https://github.com/fabioricali/timerlist/blob/master/api.md">https://github.com/fabioricali/timerlist/blob/master/api.md</a>

### Browser

#### Local
```html
<script src="node_modules/timerlist/dist/timerlist.min.js"></script>
```

#### CDN unpkg
```html
<script src="https://unpkg.com/timerlist/dist/timerlist.min.js"></script>
```

#### CDN jsDeliver
```html
<script src="https://cdn.jsdelivr.net/npm/timerlist/dist/timerlist.min.js"></script>
```

## Changelog
You can view the changelog <a target="_blank" href="https://github.com/fabioricali/timerlist/blob/master/CHANGELOG.md">here</a>

## License
TimerList is open-sourced software licensed under the <a target="_blank" href="http://opensource.org/licenses/MIT">MIT license</a>

## Author
<a target="_blank" href="http://rica.li">Fabio Ricali</a>

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [1.1.0] - 2017-11-25
- Added support to UMD

## [1.0.2] - 2017-08-15
- Fixed reamde

## [1.0.1] - 2017-08-15
- First release

test/timerlist.js
const Timerlist = require('../src/timerlist');
const be = require('bejs');

describe('timerlist', function () {
    this.timeout(5000);

    describe('setTimeout', function () {
        it('should be run after 100 ms', (done) => {

            const timer = new Timerlist();

            timer.setTimeout('myname', (param)=>{
                console.log('hello', param);
                done();
            }, 100, 'world');

        });
        it('reset timeout, should be run once', (done) => {

            const timer = new Timerlist();

            timer.setTimeout('myname', (param)=>{
                console.log('hello', param);
                done();
            }, 1000, 'world');

            timer.setTimeout('myname', (param)=>{
                console.log('hello', param);
                done();
            }, 1000, 'world');

        });
    });

    describe('setInterval', function () {
        it('should be run for 300 ms', (done) => {

            const timer = new Timerlist();
            let calls = 0;

            timer.setInterval('myname', (param)=>{
                if(++calls === 3) {
                    console.log(timer.getInterval('myname'));
                    // destroy timer
                    timer.clearInterval('myname');
                    done();
                }
                console.log(calls);
                console.log('hello', param);
            }, 100, 'world');

        });

        it('reset interval, should be run once', (done) => {

            const timer = new Timerlist();

            console.log(timer.getTimers());

            timer.setInterval('myname', (param)=>{
                console.log('hello', param);
                done();
            }, 1500, 'world');

            timer.setInterval('myname', (param)=>{
                console.log('hello', param);
                done();
            }, 1500, 'world');

        });
    });

    describe('clearTimeout', function () {
        it('undefined, should be return void', () => {

            const timer = new Timerlist();

            timer.clearTimeout('myname');

        });
        it('should be return undefined', () => {

            const timer = new Timerlist();
            timer.setTimeout('myname', ()=>{}, 2000);
            be.err.object(timer.getTimeout('myname'));
            timer.clearTimeout('myname');
            be.err.undefined(timer.getTimeout('myname'));
        });
    });

    describe('clearInterval', function () {
        it('undefined, should be return void', () => {

            const timer = new Timerlist();

            timer.clearInterval('myname');

        });
        it('should be return undefined', () => {

            const timer = new Timerlist();
            timer.setInterval('myname', ()=>{}, 2000);
            be.err.object(timer.getInterval('myname'));
            timer.clearInterval('myname');
            be.err.undefined(timer.getInterval('myname'));
        });
    });

    describe('clearAll', function () {
        it('should be return empty timers list', () => {

            const timer = new Timerlist();

            timer.setTimeout('myname1', ()=>{}, 100);
            timer.setTimeout('myname2', ()=>{}, 100);
            timer.setTimeout('myname3', ()=>{}, 100);

            timer.setInterval('myname1', ()=>{}, 100);
            timer.setInterval('myname2', ()=>{}, 100);
            timer.setInterval('myname3', ()=>{}, 100);

            let {timeout, interval} = timer.getTimers();
            be.err.not.empty(timeout);
            be.err.not.empty(interval);

            timer.clearAll();

            let {timeout1, interval1} = timer.getTimers();
            be.err.empty(timeout1);
            be.err.empty(interval1);
        });
    });

    describe('maxDelay', function () {
        it('should be return a number', ()=>{
            be.err.equal(Timerlist.maxDelay(), 2147483647);
        })
    });

    describe('isTimeoutCalled', function () {
        it('should be return true', (done)=>{
            const timer = new Timerlist();
            timer.setTimeout('myname', ()=>{}, 500);
            setTimeout(()=>{
                be.err(done).true(timer.isTimeoutCalled('myname'));
            }, 600);
        });
        it('should be return false', ()=>{
            const timer = new Timerlist();
            timer.setTimeout('myname', ()=>{}, 500);
            be.err.false(timer.isTimeoutCalled('myname'));
        });
    });

    describe('isIntervalCalled', function () {
        it('should be return true', (done)=>{
            const timer = new Timerlist();
            timer.setInterval('myname', ()=>{}, 500);
            setTimeout(()=>{
                be.err(done).true(timer.isIntervalCalled('myname'));
                timer.clearInterval('myname');
            }, 600);
        });
        it('should be return false', ()=>{
            const timer = new Timerlist();
            timer.setInterval('myname', ()=>{}, 500);
            be.err.false(timer.isIntervalCalled('myname'));
            timer.clearInterval('myname');
        });
    });

    describe('_createTimer', function () {
        it('undefined name, should be return error', (done)=>{
            const timer = new Timerlist();
            try {
                timer._createTimer({
                    type: 'timeout'
                });
            } catch (e) {
                console.log(e.message);
                be.err(done).equal(e.message, Timerlist._ERROR_MESSAGE[0]);
            }
        });

        it('undefined callback, should be return error', (done)=>{
            const timer = new Timerlist();
            try {
                timer._createTimer({
                    name: 'myname',
                    type: 'timeout'
                });
            } catch (e) {
                console.log(e.message);
                be.err(done).equal(e.message, Timerlist._ERROR_MESSAGE[1]);
            }
        });

        it('undefined delay, should be return error', (done)=>{
            const timer = new Timerlist();
            try {
                timer._createTimer({
                    name: 'myname',
                    callback: ()=>{},
                    type: 'timeout'
                });
            } catch (e) {
                console.log(e.message);
                be.err(done).equal(e.message, Timerlist._ERROR_MESSAGE[2]);
            }
        });
    });

});

webpack.config.js
const webpack = require('webpack');
const unminifiedWebpackPlugin = require('unminified-webpack-plugin');
const WebpackAutoInject = require('webpack-auto-inject-version');
const pkg = require('./package.json');

const banner = `
  ${pkg.name} - ${pkg.description}
  Author: ${pkg.author.name}
  Version: v${pkg.version}
  Url: ${pkg.homepage}
  License(s): ${pkg.license}
`;

module.exports = {
    entry: './index.js',
    output: {
        filename: './dist/timerlist.min.js',
        library: 'timerlist',
        umdNamedDefine: true,
        libraryTarget: 'umd'
    },
    resolve: {
        modules: ['node_modules'],
        extensions: ['*', '.js'],
    },
    module: {
        rules: [{
            test: /\.js?$/,
            loader: 'babel-loader',
            options: {
                presets: ['es2015'],
            },
        }],
    },
    plugins: [
        new webpack.optimize.UglifyJsPlugin({
            mangle: true,
            comments: false,
            compress: {
                warnings:false
            }, include: /\.min\.js$/ }),
        new WebpackAutoInject({
            PACKAGE_JSON_PATH: './package.json',
            components: {
                InjectAsComment: true,
                InjectByTag: true,
            },
            componentsOptions: {
                InjectAsComment: {
                    tag: 'TimerList build version: {version}'
                }
            }
        }),
        new unminifiedWebpackPlugin()/*,
        new webpack.BannerPlugin(banner)*/
    ]
};


src/timerlist.js
/**
 * @ignore
 */
const ERROR_MESSAGE = [
    'name must be a string',
    'callback must be a function',
    'delay must be a number',
];

class TimerList {

    /**
     * @constructor
     * @example
     * const TimerList = require('timerlist');
     * const timer = new TimerList();
     * timer.setTimeout('myTimer', ()=>{
     *      console.log('my timer is running');
     * }, 1000);
     */
    constructor() {
        this.timers = {
            timeout: {},
            interval: {}
        };
    }

    /**
     * Get max number delay (approximately 24.8 days)
     * @returns {number}
     */
    static maxDelay() {
        return 2147483647;
    }

    /**
     * Create timer
     * @param params {Object}
     * @private
     * @ignore
     */
    _createTimer(params) {
        if(typeof params.name !== 'string')
            throw new Error(ERROR_MESSAGE[0]);

        if(typeof params.callback !== 'function')
            throw new Error(ERROR_MESSAGE[1]);

        if(typeof params.delay !== 'number')
            throw new Error(ERROR_MESSAGE[2]);

        params.args.unshift(params.callback, params.delay);

        if(params.type === 'timeout') {

            if (this.getTimeout(params.name))
                this.clearTimeout(params.name);

            this.timers.timeout[params.name] = setTimeout.apply(this, params.args);
        } else {

            if (this.getInterval(params.name))
                this.clearInterval(params.name);

            this.timers.interval[params.name] = setInterval.apply(this, params.args);
        }
    }

    /**
     * Set timeout
     * @param name {string} timer name
     * @param callback {Function} callback function
     * @param delay=0 {number} delay interval
     * @param [args] {...args} optional arguments
     * @example
     * timer.setTimeout('my timer', myListener, 1000);
     */
    setTimeout(name, callback, delay = 0, ...args) {
        this._createTimer({
            type: 'timeout',
            name: name,
            callback: callback,
            delay: delay,
            args: args
        });
    }

    /**
     * Set interval
     * @param name {string} timer name
     * @param callback {Function} callback function
     * @param delay=0 {number} delay interval
     * @param [args] {...args} optional arguments
     * @example
     * timer.setInterval('my timer', myListener, 1000);
     */
    setInterval(name, callback, delay = 0, ...args) {
        this._createTimer({
            type: 'interval',
            name: name,
            callback: callback,
            delay: delay,
            args: args
        });
    }

    /**
     * Check if a timeout was called
     * @param name {string} timer name
     * @returns {boolean}
     */
    isTimeoutCalled(name) {
        return this.timers.timeout[name]._called;
    }

    /**
     * Check if an interval was called
     * @param name {string} timer name
     * @returns {boolean}
     */
    isIntervalCalled(name) {
        return this.timers.interval[name]._called;
    }

    /**
     * Get all timers
     * @returns {{timeout: {}, interval: {}}|*}
     */
    getTimers() {
        return this.timers;
    }

    /**
     * Destroy all timers
     */
    clearAll() {
        for(let i in this.timers.timeout){
            if(this.timers.timeout.hasOwnProperty(i))
                this.clearTimeout(this.timers.timeout[i]);
        }
        for(let i in this.timers.interval){
            if(this.timers.interval.hasOwnProperty(i))
                this.clearTimeout(this.timers.interval[i]);
        }
    }

    /**
     * Get timer timeout reference
     * @param name {string} timer name
     * @returns {*}
     */
    getTimeout(name) {
        return this.timers.timeout[name];
    }

    /**
     * Get timer interval reference
     * @param name {string} timer name
     * @returns {*}
     */
    getInterval(name) {
        return this.timers.interval[name];
    }

    /**
     * Destroy a timeout
     * @param name {string} timer name
     */
    clearTimeout(name) {
        clearTimeout(this.timers.timeout[name]);
        delete this.timers.timeout[name];
    }

    /**
     * Destroy an interval
     * @param name {string} timer name
     */
    clearInterval(name) {
        clearInterval(this.timers.interval[name]);
        delete this.timers.interval[name];
    }
}

module.exports = TimerList;
module.exports._ERROR_MESSAGE = ERROR_MESSAGE;